### DISCUSSION TITLE
Samba AD DC running in Kubernetes

### DISCUSSION THREAD

**Discussion by rsepulvedacl** [samba-in-kubernetes/sambacc#112] [Show and tell]
Hi everyone!

I'd like to show you a repository where I've been publishing a YAML that allowed me to run Samba AD DC successfully, exposing ports out from the cluster.
https://github.com/burbuja/samba-in-kubernetes

Anyways, I feel that I need to eliminate the bash script and the manual steps. To achieve that, I'd like to see these two functions in sambacc:
* A parameter to pass an external IP address (--host-ip, maybe?), this would avoid to change it manually when the pod is already running.
* Since this approach uses a stateful set to run the container, it would be useful to take advantage of that to whether provision a new domain or join to one, depending on the pod index number.

I'd like to see your comments.

**Comment by phlogistonjohn** [comment]
> I'd like to show you a repository where I've been publishing a YAML that allowed me to run Samba AD DC successfully, exposing ports out from the cluster. https://github.com/burbuja/samba-in-kubernetes
> 

That's pretty cool, thanks for sharing!


> Anyways, I feel that I need to eliminate the bash script and the manual steps. To achieve that, I'd like to see these two functions in sambacc:
> 

I didn't see a bash script in the repo... did I overlook something?


>     * A parameter to pass an external IP address (--host-ip, maybe?), this would avoid to change it manually when the pod is already running.
> 

What should `--host-ip` do? In other words, what are you currently doing to "change it manually" are you running `samba-tool` commands?


>     * Since this approach uses a stateful set to run the container, it would be useful to take advantage of that to whether provision a new domain or join to one, depending on the pod index number.
> 
> 

Yeah, I never put much effort into making multi-dc work properly. I mainly use the ad-dc image for testing purposes and these domains never exist for a long time and don't need to replicate. One idea I had once was to possibly support simple if-then conditions no the command line to trigger actions.  Hypothetically something like `--skip-if-condition=$VAR >= 1` but that's pretty abstract. Plus I don't want to go too far down the road of creating some sort of domain specific language either. We already have `--skip-if-file` on the file server cli. We could bring that over to the ad dc side. How far would that get you?

Happy to discuss more.

I also am trying to encourage folks who ask for features to contribute, as I am very busy lately. Even if you don't feel like a python expert I suggest giving it a whirl and I can coach you and give feedback.

**Comment by rsepulvedacl** [comment]
> That's pretty cool, thanks for sharing!

It's OK.

> I didn't see a bash script in the repo... did I overlook something?

The bash script is [part of a config map](https://github.com/burbuja/samba-in-kubernetes/blob/master/addc.yaml#L181).

> What should `--host-ip` do? In other words, what are you currently doing to "change it manually" are you running `samba-tool` commands?

I'd like to access to the pod from outside the cluster, using MetalLB, as load balancer, to expose the ports. MetalLB has its own IP addresses, but the pod starts assuming the internal IP address, then it creates registries on the DNS based on that internal IP. So, I'd like to force sambacc to use the external IP [I already set on MetalLB](https://github.com/burbuja/samba-in-kubernetes/blob/master/addc.yaml#L181) and [I set as part of /etc/hosts file](https://github.com/burbuja/samba-in-kubernetes/blob/master/addc.yaml#L625), from the begining. I think that would save me from doing some additional steps manually with samba-tool, adding the external IP and deleting the internal one, as explained on the [readme file](https://github.com/burbuja/samba-in-kubernetes/blob/master/README.md?plain=1#L80).

> Yeah, I never put much effort into making multi-dc work properly. I mainly use the ad-dc image for testing purposes and these domains never exist for a long time and don't need to replicate. One idea I had once was to possibly support simple if-then conditions no the command line to trigger actions. Hypothetically something like `--skip-if-condition=$VAR >= 1` but that's pretty abstract. Plus I don't want to go too far down the road of creating some sort of domain specific language either. We already have `--skip-if-file` on the file server cli. We could bring that over to the ad dc side. How far would that get you?

Oh, I understand. Anyways, the stateful set creates numbered pods, like: dc-0, dc-1, dc-2, etc., while it assigns a pod index number that [I pass as an environment variable](https://github.com/burbuja/samba-in-kubernetes/blob/master/addc.yaml#L508). Then, this number is taken by the [bash script](https://github.com/burbuja/samba-in-kubernetes/blob/master/addc.yaml#L181) to start the sambacc to provision (if POD_INDEX == 0) or to join to a domain (as "else").

> Happy to discuss more.
> 
> I also am trying to encourage folks who ask for features to contribute, as I am very busy lately. Even if you don't feel like a python expert I suggest giving it a whirl and I can coach you and give feedback.

I'm not a Python expert, but I could make some changes, forking sambacc source code to achieve what I'd like to see. Anyways, I feel that I need to know more about the plans that you may have about the modes that you'd pretend Samba AD DC to have in Kubernetes and define how to name and pass every option.

If you need some more evidences, like screenshots or logs, please don't hesitate to ask me for them.
